File: leet-code/622. Design Cirular Queue/DesignCircularQueue622_3.java
package com.example;

public class DesignCircularQueue622_3 {

    private int size;
    private int[] circularQueue;
    private int front;
    private int rear;

    public DesignCircularQueue622_3(int k) {
        // 큐의 크기를 k + 1 로 한다
        size = k + 1;
        circularQueue = new int[size];
        for (int i = 0; i < size; i++) {
            circularQueue[i] = -1;
        }
        front = 0;
        rear = 0;
    }

    public boolean enQueue(int value) {
        if (isFull()) {
            return false;
        }
        rear = (rear + 1) % size;
        circularQueue[rear] = value;
        return true;
    }

    public boolean deQueue() {
        if (isEmpty()) {
            return false;
        }
        circularQueue[front] = -1;
        front = (front + 1) % size;
        return true;
    }

    public int Front() {
        if (isEmpty()) {
            return -1;
        }
        return circularQueue[(front + 1) % size];
    }

    public int Rear() {
        if (isEmpty()) {
            return -1;
        }
        return circularQueue[rear];
    }

    public boolean isEmpty() {
        return front == rear;
    }

    public boolean isFull() {
        return (rear + 1) % size == front;
    }
}

File: leet-code/622. Design Cirular Queue/DesignCircularQueue622_3.kt
class DesignCircularQueue_3(k: Int) {
    val circularQ: IntArray = IntArray(k)
    val maxSize: Int = k
    var length: Int = 0;
    var front: Int = 0;
    var rear: Int = -1;

    fun enQueue(value: Int): Boolean {
        if (isFull()) {
            return false
        }
        length += 1
        rear = (rear + 1) % maxSize
        circularQ[rear] = value
        return true
    }

    fun deQueue(): Boolean {
        if (isEmpty()) {
            return false
        }
        length -= 1
        front = (front + 1) % maxSize
        return true
    }

    fun Front(): Int {
        if (isEmpty()) {
            return -1
        }
        return circularQ[front]
    }

    fun Rear(): Int {
        if (isEmpty()) {
            return -1
        }
        return circularQ[rear]
    }

    fun isEmpty(): Boolean {
        return length == 0;
    }

    fun isFull(): Boolean {
        return maxSize == length;
    }
}